src/server/middleware/auth.js

```diff
@@ -97,30 +97,12 @@ exports.getProvider = function (ctx) {
 };
 
 async function getUserEmailFromToken(ctx) {
-  if (!config.authServers.length) return 'anonymous';
-
-  const token = ctx.headers['x-auth-session'] || ctx.query['x-auth-session'];
-  if (!token) return 'anonymous';
-
-  for (let i = 0; i < config.authServers.length; i++) {
-    // eslint-disable-next-line no-await-in-loop
-    const res = await got(
-      `${config.authServers[i].replace(/\/$/, '')}/_session`,
-      {
-        responseType: 'json',
-        headers: {
-          cookie: token,
-        },
-        throwHttpErrors: false,
-      },
-    );
-
-    if (res.body && res.body.userCtx) {
-      return res.body.userCtx.name;
-    }
+  try {
```

Comment:
this function is probably poorly name, because it has nothing to do with `?token=` authentication.

Reply:
yes, this was something that confused me. I was not sure what this was supposed to do.

Reply:
right, but we can still leave the the previous code in for the `x-auth-session` and then have this check for the `?token=` case?

Reply:
that is, something like---but again, sorry if i'm wasting your time here!

```js
async function getUserEmailFromAuthToken(ctx) {
  try {
    const token = await ctx.state.couch.getToken(ctx.query.token);
    return token.$owner;
  } catch (err) {
    return 'anonymous';
  }
}
async function getUserEmailFromToken(ctx) {
  // Try to get ?token token and use this and not the x-auth cookie token
  const authToken = getUserEmailFromAuthToken(ctx);
  if (authToken !== 'anonymous') return authToken;

  if (!config.authServers.length) return 'anonymous';

  const token = ctx.headers['x-auth-session'] || ctx.query['x-auth-session'];
  if (!token) return 'anonymous';

  for (let i = 0; i < config.authServers.length; i++) {
    // eslint-disable-next-line no-await-in-loop
    const res = await got(
      `${config.authServers[i].replace(/\/$/, '')}/_session`,
      {
        responseType: 'json',
        headers: {
          cookie: token,
        },
        throwHttpErrors: false,
      },
    );

    if (res.body && res.body.userCtx) {
      return res.body.userCtx.name;
    }
  }

  return 'anonymous';
}
```

Reply:
Please revert this part of the change. The bug is not here and it's not how it should be fixed. An "entry" token only gives the right to a specific entry. If you override the user email with the one that's in the token, you effectively allow that request to access everything that the user can access (not only the entry specified by the token, and also not only with the rights specified in the token).

Reply:
The token lookup is already done for all requests at:
https://github.com/cheminfo/rest-on-couch/blob/master/src/server/routes/api.js#L21